libraries/stdlib/test/text/ParsePrimitivesJVMTest.kt

```diff
@@ -13,79 +13,96 @@ class ParsePrimitivesJVMTest {
-        assertEqualsOrFailsNullable(77.0, "77.0", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(-1.77, "-1.77", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(0.77, "+.77", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(-77.0, "\t-77 \n", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(77.0, "7.7e1", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(77.0, "+770e-1", String::toDouble, String::toDoubleOrNull)
-
-        assertEqualsOrFailsNullable(-Double.NaN, "-NaN", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(Double.POSITIVE_INFINITY, "+Infinity", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable((0x77 shl 1).toDouble(), "0x77p1", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(0x77.toDouble(), "0x.77P8", String::toDouble, String::toDoubleOrNull)
-
-        assertEqualsOrFailsNullable(null, "7..7", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(null, "0x77e1", String::toDouble, String::toDoubleOrNull)
-        assertEqualsOrFailsNullable(null, "007 not a number", String::toDouble, String::toDoubleOrNull)
+        CompareBehaviourContext(String::toDouble, String::toDoubleOrNull).apply {
+
+            assertProduce(-77.0, "-77")
+            assertProduce(77.0, "77.")
+            assertProduce(77.0, "77.0")
+            assertProduce(-1.77, "-1.77")
+            assertProduce(0.77, "+.77")
+            assertProduce(-77.0, "\t-77 \n")
+            assertProduce(77.0, "7.7e1")
+            assertProduce(77.0, "+770e-1")
+
+            assertProduce(-Double.NaN, "-NaN")
+            assertProduce(Double.POSITIVE_INFINITY, "+Infinity")
+            assertProduce((0x77 shl 1).toDouble(), "0x77p1")
+            assertProduce(0x77.toDouble(), "0x.77P8")
+
+            assertProduce(null, "7..7")
+            assertProduce(null, "0x77e1")
+            assertProduce(null, "007 not a number")
+        }
     }
 }
 
-private inline fun <T: Any> assertEqualsOrFailsNullable(output: T?,
-                                                        input: String,
-                                                        crossinline converOrFail: (String) -> T,
-                                                        crossinline convertOrNull: (String) -> T?) {
-    if(output == null) {
-        assertFails { converOrFail(input) }
-        assertNull (convertOrNull(input) )
-    } else {
-        assertEquals(output, converOrFail(input))
-        assertEquals(output, convertOrNull(input))
+
+private class CompareBehaviourContext<T: Any>(val convertOrFail: (String) -> T,
+                                              val convertOrNull: (String) -> T?) {
+    fun assertProduce(output: T?, input: String) {
+        if(output == null) {
+            assertFails { convertOrFail(input) }
```

Comment:
Please specify input string in the detail message, i.e. something like "Expected to fail (or return null) on $input"